### PR TITLE
Don't return replies to hidden annotations in the bulk API

### PR DESCRIPTION
**Pull request by marcospri**
Potential fix for: https://github.com/hypothesis/support/issues/66

Join on the main annotation table to find the reference to the parent, if any and check that it's visible.

### CODE REVIEW

File: h/services/bulk_api/annotation.py [hunk 1]
```diff
@@ -74,6 +85,11 @@ def _search_query(cls, authority, username, created) -> Select:
             .select_from(AnnotationSlim)
             .join(cls._AUTHOR, cls._AUTHOR.id == AnnotationSlim.user_id)
             .join(Group, Group.id == AnnotationSlim.group_id)
+            .join(cls._ANNOTATION)
```
Comment (marcospri) [inline h/services/bulk_api/annotation.py:88]: AnnotationSlim always has a join with Annotation.

File: h/services/bulk_api/annotation.py [hunk 2]
```diff
@@ -74,6 +85,11 @@ def _search_query(cls, authority, username, created) -> Select:
...
+            .outerjoin(
```
Comment (marcospri) [inline h/services/bulk_api/annotation.py:89]: Outer join here, not all annotations have parents.

File: h/services/bulk_api/annotation.py [hunk 3]
```diff
@@ -82,6 +98,10 @@ def _search_query(cls, authority, username, created) -> Select:
                 cls._AUTHOR.nipsa.is_(False),
                 AnnotationSlim.moderated.is_(False),
                 Group.id.in_(cls._audience_groups_subquery(authority, username)),
+                or_(
+                    cls._ANNOTATION_PARENT.id.is_(None),
```
Comment (marcospri) [inline h/services/bulk_api/annotation.py:102]: Either it doesn't have a parent.

File: h/services/bulk_api/annotation.py [hunk 4]
```diff
@@ -82,6 +98,10 @@ def _search_query(cls, authority, username, created) -> Select:
...
+                    cls._ANNOTATION_PARENT.shared.is_(True),
```
Comment (marcospri) [inline h/services/bulk_api/annotation.py:103]: Or the parent is visible.

File: tests/common/factories/annotation_slim.py [hunk 5]
```diff
@@ -13,7 +13,9 @@ class AnnotationSlim(ModelFactory):
     class Meta:
         model = models.AnnotationSlim
 
-    annotation = factory.SubFactory(Annotation)
+    annotation = factory.SubFactory(
```
Comment (marcospri) [inline tests/common/factories/annotation_slim.py:16]: Use the shared from the Slim factory into the Annotation one.

File: h/services/bulk_api/annotation.py [hunk 6]
```diff
@@ -74,6 +85,11 @@ def _search_query(cls, authority, username, created) -> Select:
             .select_from(AnnotationSlim)
             .join(cls._AUTHOR, cls._AUTHOR.id == AnnotationSlim.user_id)
             .join(Group, Group.id == AnnotationSlim.group_id)
+            .join(cls._ANNOTATION)
+            .outerjoin(
+                cls._ANNOTATION_PARENT,
+                cls._ANNOTATION.references[0] == cls._ANNOTATION_PARENT.id,
```
Comment (seanh) [inline h/services/bulk_api/annotation.py:91]: I'm guessing `references[0]` is the root annotation of the thread? What if a reply in the middle of a thread is hidden, and there are replies to that hidden reply?

**Review by seanh — COMMENTED** [review]
I wonder if we might want to go in the opposite direction when fixing <https://github.com/hypothesis/support/issues/66>. After all, the student selected `Post to <GROUP>` when posting the annotation, so the group should be able to see it. Perhaps we should fix the client to be able to show shared replies to only-me annotations (or more generally: to be able to show replies when the parent annotation(s) can't be shown). We may have to show them in the orphans tab. I wouldn't be surprised if there was already a client issue for this somewhere...

**Comment by robertknight** [comment]
>  Perhaps we should fix the client to be able to show shared replies to only-me annotations (or more generally: to be able to show replies when the parent annotation(s) can't be shown).

The client can already show replies where the parent annotation is missing, if it receives them.

**Comment by marcospri** [comment]
Closing this in favour of a more general solution, ie one that's not scoped only to email digest.

See: https://hypothes-is.slack.com/archives/C1MA4E9B9/p1719566013178539

**Pull request closed without merging**